### PR TITLE
Enable [Environment]::OSVersion to return current OS rather than compatible version

### LINKED ISSUE #6452: Powershell Core Reports the wrong version of windows when referencing c# Assemblies
I apologize if this issue already exists and I missed it or if you are already aware of this issue. This is my first time submitting an issue here.

When using the System.Environment assembly to get the Windows OS version that is it running PowerShell reports the same version of Windows as Windows 8/Windows Server 2012 instead of the correct version of Windows on Windows 8.1/Windows Server 2012R2 and Windows 10/Windows Server 2016. I found a [stackoverflow](https://stackoverflow.com/questions/2819934/detect-windows-version-in-net) article referencing [Windows Developer document](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724833(v=vs.85).aspx). It is probably not this simple but you might need to [modify your manifest file](https://msdn.microsoft.com/en-us/library/windows/desktop/dn481241(v=vs.85).aspx) for your PowerShell builds.

Steps to reproduce
------------------

```powershell
[System.Environment]::OSVersion.Version

```

Expected behavior
-----------------

```none
[System.Environment]::OSVersion.Version

Major  Minor  Build  Revision
-----  -----  -----  --------
10     0      16299  0
```

Actual behavior
---------------

```none
[System.Environment]::OSVersion.Version

Major  Minor  Build  Revision
-----  -----  -----  --------
6      2      9200   0
```

Environment data
----------------

```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.2
PSEdition                      Core
GitCommitId                    v6.0.2
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

Without compatibility flags set in an application manifest, dotnet will report the OS as the compatible version instead of the actual OS version.
Fix is to add an application manifest that includes the <compatibility> section indicating compatibility.
Also added tests for existing resources we write into pwsh.exe.

Fix https://github.com/PowerShell/PowerShell/issues/6452

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by daxian-dbw — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 1]
```diff
@@ -663,3 +663,18 @@ Describe "Console host api tests" -Tag CI {
...
+        $pwsh = Get-Item -Path "$PSHOME\pwsh.exe"
+        $pwsh.VersionInfo.FileVersion | Should -BeExactly $PSVersionTable.PSVersion
+        $pwsh.VersionInfo.ProductVersion.Replace("-dirty","") | Should -BeExactly $PSVersionTable.GitCommitId.Replace("v","")
+        $pwsh.VersionInfo.ProductName | Should -BeExactly "PowerShell Core 6"
+    }
+
+    It "Manifest contains compatibility section" -Skip:(!$IsWindows) {
+        $osversion = [System.Environment]::OSVersion.Version
+        $osversionString = $psversiontable.os -replace "^Microsoft Windows (\d+\.\d+).*$",'$1'
+        "$($osversion.Major).$($osversion.Minor)" | Should -BeExactly $osversionString
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: Maybe use Should -Match
```powershell
$psversiontable.os | Should -Match "$($osversion.Major).$($osversion.Minor)" 
```

**Comment by adityapatwardhan** [comment]
@iSazonov are you comments blocking signoff?

**Comment by iSazonov** [comment]
I don't understand why we have `pwsh.1.manifest` in the assert directory and use `pwsh.manifest` name in the script&

File: build.psm1 [hunk 2]
```diff
@@ -630,7 +630,8 @@ Fix steps:
 
         Start-NativeExecution { & "~/.rcedit/rcedit-x64.exe" $pwshPath --set-icon "$PSScriptRoot\assets\Powershell_black.ico" `
             --set-file-version $fileVersion --set-product-version $ReleaseVersion --set-version-string "ProductName" "PowerShell Core 6" `
-            --set-requested-execution-level "asInvoker" --set-version-string "LegalCopyright" "(C) Microsoft Corporation.  All Rights Reserved." } | Write-Verbose
+            --set-version-string "LegalCopyright" "(C) Microsoft Corporation.  All Rights Reserved." `
+            --application-manifest "$PSScriptRoot\assets\pwsh.manifest" } | Write-Verbose
```
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply build.psm1]: Will fix

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 1 above)
Comment (SteveL-MSFT) [reply test/powershell/Host/ConsoleHost.Tests.ps1]: This is simpler, will change

File: build.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (SteveL-MSFT) [reply build.psm1]: surprised this wasn't caught in CI unless `rcedit` tries looking for `pwsh.1.manifest`

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 3]
```diff
@@ -663,3 +663,17 @@ Describe "Console host api tests" -Tag CI {
...
+    It "Resource strings are embedded in the executable" -Skip:(!$IsWindows) {
+        $pwsh = Get-Item -Path "$PSHOME\pwsh.exe"
+        $pwsh.VersionInfo.FileVersion | Should -BeExactly $PSVersionTable.PSVersion
+        $pwsh.VersionInfo.ProductVersion.Replace("-dirty","") | Should -BeExactly $PSVersionTable.GitCommitId.Replace("v","")
+        $pwsh.VersionInfo.ProductName | Should -BeExactly "PowerShell Core 6"
+    }
+
+    It "Manifest contains compatibility section" -Skip:(!$IsWindows) {
+        $osversion = [System.Environment]::OSVersion.Version
+        $psversiontable.os | Should -Match "$($osversion.Major).$($osversion.Minor)"
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: @SteveL-MSFT Please clarify - if we usually use `-BeExactly` for string should we use `-MatchExactly` for string too?

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT  Please have a look at CI failure:

`Expected exactly 6.1.0-preview.8660, but got '6.1.0'.`

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 3 above)
Comment (SteveL-MSFT) [reply test/powershell/Host/ConsoleHost.Tests.ps1]: Makes sense, will change

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 4]
```diff
@@ -667,13 +667,17 @@ Describe "Console host api tests" -Tag CI {
 Describe "Pwsh exe resources tests" -Tag CI {
     It "Resource strings are embedded in the executable" -Skip:(!$IsWindows) {
         $pwsh = Get-Item -Path "$PSHOME\pwsh.exe"
-        $pwsh.VersionInfo.FileVersion | Should -BeExactly $PSVersionTable.PSVersion
+        $fileVersion = $PSVersionTable.PSVersion.ToString()
+        if ($fileVersion.Contains("-")) {
+            $fileVersion = $fileVersion.Split("-")[0]
+        }
```
Comment (daxian-dbw) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: @SteveL-MSFT I'm curious why this PR causes this test to fail? Can you please help me understand?

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 5]
```diff
@@ -667,13 +667,17 @@ Describe "Console host api tests" -Tag CI {
 Describe "Pwsh exe resources tests" -Tag CI {
     It "Resource strings are embedded in the executable" -Skip:(!$IsWindows) {
         $pwsh = Get-Item -Path "$PSHOME\pwsh.exe"
-        $pwsh.VersionInfo.FileVersion | Should -BeExactly $PSVersionTable.PSVersion
+        $fileVersion = $PSVersionTable.PSVersion.ToString()
+        if ($fileVersion.Contains("-")) {
```
Comment (iSazonov) [inline test/powershell/Host/ConsoleHost.Tests.ps1]: Seems we can remove the `if`.

Comment (SteveL-MSFT) [reply test/powershell/Host/ConsoleHost.Tests.ps1]: You're right.  Will update.

**Review by iSazonov — APPROVED** [review]

**Pull request merged**